### ISSUE TITLE
Overly verbose section permalinks for screen reader users

### ISSUE THREAD

**Issue by jscholes** [w3c/respec#3898]
## Important info

- URL to affected spec: https://w3c.github.io/aria-practices/
- ReSpec version: Unknown; there is no ReSpec button on the linked editor's draft.
- [x] I did a "hard refresh", but it's still busted.

## Description of problem

As highlighted in https://github.com/w3c/aria-practices/issues/2159, the changes in https://github.com/w3c/respec/pull/3874 make all section permalinks, and hence section headings, incredibly verbose for screen reader users.  That issue expands on the problematic behaviour and a suggested fix.

CC @marcoscaceres / @mcking65 / @sinabahram

**Comment by jscholes** [comment]
@jnurthen Thanks for the quick patch.  As the PR doesn't fully resolve this issue, can it be reopened?

**Comment by marcoscaceres** [comment]
@jscholes, reopened for you... can you provide me with some additional guidance on how this should work on the ReSpec side?

**Comment by jscholes** [comment]
@marcoscaceres Certainly!  Instead of having a heading with both the text and a labelled permalink, the text of the heading itself should be the permalink.  E.g.:

```
<h2 id="x3-design-patterns-and-widgets">
	<a class="self-link" href="#aria_ex">
		<bdi class="secno">3. </bdi>
		Design Patterns and Widgets
	</a>
</h2>
```

**Comment by marcoscaceres** [comment]
Ok. So you are suggesting removing `aria-label` entirely, so (current):

```HTML
<h2 id="x1-introduction">
  <bdi class="secno">1. </bdi>Introduction
   <a class="self-link" href="#intro" aria-label="Permalink for Section 1."></a>
</h2>
```

Becomes:

```HTML
<h2 id="x1-introduction">
  <a class="self-link" href="#x1-introduction"><bdi class="secno">1. </bdi>Introduction</a>
</h2>
```

And we would need to remove the underline from the links (and continue to add the section marker via CSS). 

@jnurthen, wdyt?

**Comment by jnurthen** [comment]
@marcoscaceres I defer to screen reader users on this and what works best for them.

**Comment by sidvishnoi** [comment]
There is chance of a heading already containing a link, so wrapping it in a self link is risk prone. We can probably find existing examples of this via reffy.

If we want to use wrapped self-link approach (which I also prefer), we should discourage the use of heading containing links by emitting a ReSpec warning, or maybe in those cases fallback to current section marker approach.

**Comment by pkra** [comment]
Amber Wilson's article from last year might be of interest: https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/ 

The solution puts the link outside the heading (to avoid noise), e.g.,

```html
<div class="heading-wrapper">
    <h2 id="automating-accessible-anchor-links">Automating accessible anchor links</h2>
    <a href="#automating-accessible-anchor-links">
        <span class="anchor-icon" aria-hidden="true"><svg>...</svg></span>
        <span class="hidden">Section titled Automating accessible anchor links</span>
    </a>
</div>
```

**Comment by jscholes** [comment]
I usually like to cut down on duplication and verbosity as much as possible.  So if it can be guaranteed that headings won't contain links, or there is a fallback in place for when they do, I'd still vote for the heading text to be marked up as the permalink as the primary method.  Having said that:

1. The approach used across the same document should be consistent.  That is, if there is a fallback for headings containing links, and at least one heading in a single document does, all headings should use the fallback.
2. Given the length of many W3C documents, and the likelihood that many users will be navigating by heading just as much as (if not more than) via a sequential reading order, having the permalinks after the headings is an acceptable fallback.
3. When rendering the fallback links after the headings, the linked example uses text such as, "Section titled Automating accessible anchor links".  The "Section titled ..." prefix will become obnoxiously repetitive; just the section number and name are fine (maybe with a "(permalink)" suffix: "1. Introduction (permalink)".

**Comment by pkra** [comment]
Apologies if the text content of the example side tracked the discussion. I only wanted to share another option for markup, one which had received positive feedback from the community.

I'm wondering if  changing the aria-label of the current link markup to just "permalink" (or something similar) would be an option? That would be less verbose and clear enough in context (i.e., meet wcag 2.4.4).

**Comment by marcoscaceres** [comment]
Just coming back to this. I ran into an issue with Amber Wilson's solution. 

Axe (the accessibility validator that ReSpec uses) complains that because the `<span>` is hidden with CSS (`display: none`), it won't be visible to a screen reader: It complains that, "Links must have discernible text.".

So, I think Amber Wilson's "solution 4" is our best option here, which is to use `aria-label`. I tested that with VoiceOver and works quite well.
